### DISCUSSION TITLE
JOSS paper

### DISCUSSION THREAD

**Discussion by navidcy** [FourierFlows/GeophysicalFlows.jl#206] [General]
Hey co-contributors,

I have a first draft of the JOSS paper in [JOSS-paper](https://github.com/FourierFlows/GeophysicalFlows.jl/tree/JOSS-paper) branch. The compiled pdf can be downloaded from:

http://res.cloudinary.com/hju22ue2k/image/upload/v1611563203/emcv5133bs7h8hiw5yzs.pdf

Have a look -- any suggestions welcome! I'm not saying it's ready to go but it's perhaps a good time to hear what everybody thinks? Feel free to edit and submit a PR. Check your affiliations... If you edit and you wanna see the compiled version of the changes you made then you can do so here: https://whedon.theoj.org

cc @glwagner, @liasiegelman, @BrodiePearson, @apaloczy

(P.S.: @apaloczy I don't recall hearing back from you. I'm very excited to include you in the co-authors list, but if I don't hear anything back from you the time of submission I'll assume you don't wanna be part of it.)

**Comment by apaloczy** [comment]
Hi everyone,

Thank you very much for including me in the discussion, I've just read the
JOSS draft and would be very excited to be part of it if you think I can
contribute. Just to clarify, the idea is to have this JOSS paper be a short
note on the capabilities of GeophysicalFlows.jl, correct? Would we want to
include just this one example of the Eady problem or do we have room for
maybe two or three examples?

**Comment by navidcy** [comment]
JOSS papers are usually very short (1-2 pages); e.g. https://joss.theoj.org/papers/10.21105/joss.02018 or https://joss.theoj.org/papers/10.21105/joss.01043). Some have some more examples. But note that as part of the review process the reviewers look at the package's documentation. There are lots more of examples there!

**Comment by BrodiePearson** [comment]
Thanks for pulling this together Navid! I made some suggestions and submitted a PR #189  with the [pdf here](http://res.cloudinary.com/hju22ue2k/image/upload/v1611703165/agfkyufu9m6srokr5w2b.pdf).

I haven't submitted to JOSS before, so I hope I didn't add too much...

**Comment by navidcy** [comment]
Thanks all.

Some questions for you to ponder on:

- Should we add anything regarding performance? Perhaps point to the (bit outdated) https://github.com/FourierFlows/BenchFourierFlows.jl? Personally I'd rather refrain from doing that because we enter into the loop hole of how do you compare two codes on equal basis, some are MPI-parallelized, some are multi-threaded, GPU-enabled, etc...

- Does the last paragraph reads a bit out of place? I added it because [JOSS](https://joss.readthedocs.io/en/latest/submitting.html) mentions that the paper should include "A list of key references, including to other software addressing related needs." Could we smooth it a bit you reckon?

- I'm not sure about the last paragraph regarding "ongoing research etc". There are a few other ongoing research efforts using `GeophysicalFlows.jl`... Should we add all them? Or drop everything whatsoever and perhaps we should just mention everythng to the editor upon submission?

**Comment by BrodiePearson** [comment]
- You're right that boiling down performance benchmarks is hard. I don't mind if we leave it out
- For the second-to-last paragraph, perhaps rephrasing the first sentence will help make it feel more connected: 

> GeophysicalFlows.jl is a unique Julia package, and it has similar functionality to the Python package pyqg (Abernathey et al., 2019).

- For the final paragraph, I think that mentioning the breadth of active research projects is important to demonstrate the utility of the package for research (I anticipated everyone would expand to include their relevant projects). Should we add a sentence about its utility for teaching as well? 

> GeophysicalFlows.jl can be used to investigate a variety of scientific research questions thanks to its various modules and high customizability, and its ease-of-use makes it an ideal teaching tool for fluids courses.

**Comment by navidcy** [comment]
Yes, I feel that a mention to teaching, perhaps pointing, e.g., to [this notebook](https://nbviewer.jupyter.org/github/FourierFlows/GeophysicalFlows-Examples/blob/master/RossbyWaveDispersion/RossbyWavepacket.ipynb) (?), is very good point to be made.

**Comment by liasiegelman** [comment]
Thanks for including me here. 

I agree with @BrodiePearson . In the last paragraph, you write "Dedalus (Burns et al., 2020) is Python software with an intuitive script-based interface that uses spectral methods to solve general partial differential equations, including the ones withing GeophysicalFlows.jl. " I assume you mean that Dedalus solves the same equations as GeophysicalFlows.jl and not that GeophysicalFlows.jl calls Dedalus. If I am correct, you may write something more compact like "GeophysicalFlows.jl is a unique Julia package, and it has similar functionality to the Python package pyqg (Abernathey et al., 2019) and the Python spectral solver Dedalus (Burns et al., 2020). "

Otherwise, happy to add a line or two to describe the SingleLayerQG if needed although it's already clearly described in the docs.

**Comment by navidcy** [comment]
@liasiegelman thanks! I'll rephrase.

I think we shouldn't include module-specific descriptions in the paper for brevity; the docs describe what each module does very nicely I feel.

**Comment by navidcy** [comment]
@liasiegelman, perhaps you could add a line on what you're up to with `SingleLayerQG` and Jupiter in the "ongoing research" paragraphs? You can leave it as vague as you wish...

**Comment by liasiegelman** [comment]
@navidcy ok, will do

**Comment by navidcy** [comment]
I added what was discussed above. See latest version [here](http://res.cloudinary.com/hju22ue2k/image/upload/v1611877200/hbtdd0nkhxbcmc5oypij.pdf).

What about mentioning something like "`GeophysicalFlows.jl` is continuously tested..."?

**Comment by BrodiePearson** [comment]
Do you mean there should be a new paragraph about testing?

Also, could you add in the acknowledgements:

> BCP was supported by the National Science Foundation under Grant No. 2023721

**Comment by navidcy** [comment]
Was mostly thinking a sentence about testing... or half a sentence.

I'm adding the acknowledgment now.

**Comment by navidcy** [comment]
@liasiegelman, regarding your remark:

> In the last paragraph, you write "Dedalus (Burns et al., 2020) is Python software with an intuitive script-based interface that uses spectral methods to solve general partial differential equations, including the ones withing GeophysicalFlows.jl. " I assume you mean that Dedalus solves the same equations as GeophysicalFlows.jl and not that GeophysicalFlows.jl calls Dedalus. If I am correct, you may write something more compact like "GeophysicalFlows.jl is a unique Julia package, and it has similar functionality to the Python package pyqg (Abernathey et al., 2019) and the Python spectral solver Dedalus (Burns et al., 2020)."

I rephrased here:
https://github.com/FourierFlows/GeophysicalFlows.jl/commit/2f5f2345fc3f1e9dcacc2d18bb29af865b6e8ca5

Clearer?

**Comment by liasiegelman** [comment]
@navidcy yep:)

**Comment by navidcy** [comment]
Latest version @ http://res.cloudinary.com/hju22ue2k/image/upload/v1611880691/mlfipst5ecujyiux50ig.pdf

Please scrutinize! I'll submit as soon as I hear "OK" from all of you...

**Comment by navidcy** [comment]
Is it "**on** periodic domains" or "**in** periodic domains"? At the moment both of these appear in the manuscript...

**Comment by liasiegelman** [comment]
@navidcy could you add this line at the end of the last paragraph of the paper :
"(iv) to study the genesis and persistence of the polygons of vortices present at Jovian high latitudes (Siegelman, Young and Ingersoll, in prep)." ?
My complete affiliation is Scripps Institution of Oceanography, University of California San Diego. 
Otherwise, looks good to me :)

**Comment by navidcy** [comment]
thanks @liasiegelman; done.

I kept only the institutions in the affiliations as I understood that this is JOSS style. I'll investigate this and if so I'll add Scripps, and EAPS (for Greg) and everybody's departments, etc.

**Comment by apaloczy** [comment]
@navidcy, regarding

> "Is it "on periodic domains" or "in periodic domains"? At the moment both of these appear in the manuscript..."

perhaps "**in** periodic domains" because 1D, 2D and 3D grids are supported?

Also, what does everyone think about rephrasing the title from

**GeophysicalFlows.jl: geophysical fluid dynamics-problems solvers in periodic domains on CPUs and GPUs**

to

**GeophysicalFlows.jl: Solvers for geophysical fluid dynamics problems in periodic domains on CPUs and GPUs**

**Comment by navidcy** [comment]
thanks @apaloczy. Your title suggestion reads better; the only problem it doesn't fit in two lines and not so pleasing for the eye so I changed "CPUs and GPUs" to "CPUs & GPUs" :)

**Comment by navidcy** [comment]
http://res.cloudinary.com/hju22ue2k/image/upload/v1611953748/y9zip8e6j4xcgqz594fc.pdf

**Comment by navidcy** [comment]
@BrodiePearson OK to submit?

**Comment by BrodiePearson** [comment]
@navidcy Could you re-add my funding acknowledgement in? For the last paragraph, should point (iii) be Kolmogorov?

Other than that, the new title sounds good and I'm on board for submitting!

**Comment by navidcy** [comment]
Damn! The grant acknowledgment went out by accident --- back in now.

"Kolmogorov flow" is sort of standard nomenclature in maths community about 2D Navier-Stokes with constant sinusoidal forcing; see e.g., https://aip.scitation.org/doi/10.1063/5.0002751

**Comment by BrodiePearson** [comment]
Thanks! I meant the PDF I saw says Kologorov instead of Kolmogorov...maybe it got updated after

**Comment by navidcy** [comment]
Thanks all. I clicked the "submit" button. Will let you know what happens..

**Comment by navidcy** [comment]
The submission's pre-review process is at https://github.com/openjournals/joss-reviews/issues/2996
Review is done at https://github.com/openjournals/joss-reviews/issues/3053

**Comment by navidcy** [comment]
Hey all,

I made an attempt to restructure the paper to address the [JOSS review remarks](https://github.com/openjournals/joss-reviews/issues/3053). I pushed the paper in the `JOSS-paper` branch; a preview of its current form is [here](http://res.cloudinary.com/hju22ue2k/image/upload/v1617077026/hobde1gzdtmxrba9sof2.pdf). I'm sure it can be improved....

Could you all please have a read. Let me know of any suggestions you may have. In particular, I'd appreciate

- Feedback on the remarks on `MAOOAM`, `qgs`, `Oceananigans.jl` that I've included,
- Making sure I've addressed all comments by the reviewers. (At least those comments that required something to be added in the paper!) For reviewer 1 have a look at [this post](https://github.com/openjournals/joss-reviews/issues/3053#issuecomment-790762419) and also [this one](https://github.com/openjournals/joss-reviews/issues/3053#issuecomment-792486483); for reviewer 2 have a look [here](https://github.com/openjournals/joss-reviews/issues/3053#issuecomment-796390180)

There is still the remark about GPU and scalability; we should add a remark on that in the README and in the Docs. Will do that soon...

cc: @apaloczy @BrodiePearson @glwagner @liasiegelman

**Reply by BrodiePearson** [reply]
@navidcy Thank you for addressing virtually all of these comments already! The restructuring and elaboration works well. I have a few suggestions for the revised manuscript, I hope they help:
- Line 15: `...understanding geophysical and astrophysical systems, particularly the atmospheres and oceans of Earth and other planets.`
- Line 18: `...the same partial differential equations for research or classwork. The GeophysicalFlows.jl package is designed to be easily utilized and adaptable for a wide variety of research and pedagogical purposes.`
- Line 47: `is a` (missing a)
- Line 50: [hopefully @glwagner has a more accurate and precise comparison, but this is a rough attempt]`Oceananigans.jl is an incompressible fluid solver that can be run in 1-D, 2-D, and 3-D domains. Both Oceananigans.jl and GeophysicalFlows.jl contain modules for two-dimensional flow, but the other examples provided with Oceananigans focus on three-dimensional flows which are generally present at smaller scales than the dynamics of GeophysicalFlows.jl's other modules.`
- Line 54: `The MOAAM (De Cruz et al., 2016) package, and its expanded Python implementation qgs (Demaeyer et al., 2020), simulate two atmospheric layers with QG dynamics, above either land or an oceanic layer with reduced-gravity QG dynamics. The dynamics of individual layers have overlap with the MultiLayerQG and SingleLayerQG modules, however the layer configuration of MOAAM and qgs is specifically designed to study the dynamics of Earth's mid-latitude atmosphere.`
- State of the Field section: It might be nice to bullet each model that is discussed here, [similar to the same section in the qgs paper](https://joss.theoj.org/papers/10.21105/joss.02597).
- Line 68 - The Pearson et al paper is now in press, so it can be changed to Pearson et al., 2021

**Reply by navidcy** [reply]
@BrodiePearson can you update the bib file for your paper?

**Reply by navidcy** [reply]
Thanks @BrodiePearson!
I've pushed [a commit](https://github.com/FourierFlows/GeophysicalFlows.jl/commit/19ada9ea7769532af39673edc57f9127e866a4c8) with most of your suggestions (except the one of doing bullet points for each model we compare... It's not that I don't agree with that, I just thought I sleep over it and see what others feel also before doing it.)

**Reply by BrodiePearson** [reply]
@navidcy Great, thanks for taking those on board!

**Comment by navidcy** [comment]
Latest version of the paper can be viewed [here](http://res.cloudinary.com/hju22ue2k/image/upload/v1617327417/lr9pck82dzetuccv3dus.pdf).

**Reply by liasiegelman** [reply]
Hi all, Thanks @navidcy Navid for addressing all the comments. The latest version looks good to me ! I like @BrodiePearson's suggestion of doing bullet points for each model, it would enhance readability. Otherwise, do you want to add a sentence in the paper stating that FourierFlows.jl is restricted to either a single GPU or a single CPU node atm ?

**Reply by apaloczy** [reply]
Thanks @navidcy and everyone for addressing the comments. I also feel like the bullet point layout in the comparisons to the other models could enhance readability.

This looks like a minor typo:

* Line 71: Study the bifurcation properties ```of``` Kolmogorov flows.

Also, perhaps it is worth adding to the Figure caption or to each panel's title what is the variable being plotted (PV in each layer?).

Otherwise looks good to me!

**Reply by glwagner** [reply]
I love it! My only minor comment is that "stepping stone" appears for the second time on line 35. Perhaps "starting point" could replace this second appearance of the phrase.

**Reply by navidcy** [reply]
Thanks all! 🙏 I’ll address everything and resubmit. I think I won’t mention about multiple gpus/cpus in the paper (since this is a FourierFlows.jl-related restrictions). I’ll add a remark in both FourierFlows and GeophysicalFlows docs+Readme though.

**Comment by navidcy** [comment]
Thanks all. I pushed the revised version to [JOSS review issue](https://github.com/openjournals/joss-reviews/issues/3053#issuecomment-813815364).

**Comment by navidcy** [comment]
Paper is published: https://doi.org/10.21105/joss.03053
